ruby/tests/basic_proto2.rb

```diff
@@ -43,6 +43,15 @@ def test_has_field
 
       m = TestMessage.new(:optional_int32 => nil)
       refute m.has_optional_int32?
+      refute TestMessage.descriptor.lookup('optional_int32').has?(m)
+
+      m = TestMessage.new(:optional_int32 => 0)
+      assert m.has_optional_int32?
+      assert TestMessage.descriptor.lookup('optional_int32').has?(m)
+
+      m = TestMessage.decode(TestMessage.encode(m))
+      assert m.has_optional_int32?
+      assert TestMessage.descriptor.lookup('optional_int32').has?(m)
```

Comment:
@JasonLunn Regression test added, specifically, the highlighted lines here are failing on without the patch.